lib/buildkite/exe/artifacts-csv.hs

```diff
@@ -0,0 +1,242 @@
+    , long
+    , metavar
+    , option
+    , optional
+    , progDesc
+    , strOption
+    )
+import Servant.Client
+    ( BaseUrl (..)
+    , ClientEnv
+    , ClientError
+    , ClientM
+    , Scheme (..)
+    , mkClientEnv
+    , runClientM
+    )
+import System.Environment
+    ( getEnv
+    )
+
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Text as T
+import qualified Streaming.Prelude as S
+
+organizationSlug :: Text
+organizationSlug = "cardano-foundation"
+
+mainPipeline :: PipelineName
+mainPipeline = PipelineName "cardano-wallet"
+
+adminPipeline :: PipelineName
+adminPipeline = PipelineName "cardano-wallet-buildkite-admin"
+
+buildkiteDomain :: String
+buildkiteDomain = "api.buildkite.com"
+
+buildkitePort :: Int
+buildkitePort = 443
+
+newtype PipelineName = PipelineName Text
+
+newtype Token = Token String
+
+withAuthWallet
+    :: PipelineName
+    -> LimitsLock
+    -> Token
+    -> WithAuthPipeline a
+    -> a
+withAuthWallet (PipelineName slug) l (Token apiToken) f =
+    f l (Just $ T.pack apiToken) organizationSlug slug
+
+getToken :: IO Token
+getToken = do
+    token <- getEnv "BUILDKITE_API_TOKEN"
+    pure $ Token $ "Bearer " ++ token
+
+data Options = Options
+    { _optUntilBuild :: Maybe Int
+    , _optOutputPath :: FilePath
+    }
+
+parseOptUntilBuild :: Parser (Maybe Int)
+parseOptUntilBuild =
+    optional
+        $ option auto
+        $ long "until"
+            <> metavar "BUILD_NUMBER"
+            <> help "The build number until which to collect artifacts"
+
+parseOptOutputPath :: Parser FilePath
+parseOptOutputPath =
+    strOption
+        $ long "output"
+            <> metavar "OUTPUT_PATH"
+            <> help "The path to write the CSV output to"
+
+optionsParser :: ParserInfo Options
+optionsParser =
+    info
+        ( Options
+            <$> parseOptUntilBuild
+            <*> parseOptOutputPath
+        )
+        ( fullDesc
+            <> progDesc "Collect and process benchmark history data."
+            <> header "benchmark-history - a tool for benchmark data analysis"
+        )
+
+type HandleClientError a = IO (Either ClientError a) -> IO (Maybe a)
+
+main :: IO ()
+main = do
+    bkToken <- getToken
+    Options mUntilBuild outputPath <- execParser optionsParser
+    manager <- newManager $ specialSettings False
+    lock <- newLimitsLock (LimitsLockLog >$< logsTracer) 10
+    let env = buildkiteEnv manager
+        runQuery :: HandleClientError a -> ClientM a -> IO (Maybe a)
+        runQuery f action = f $ runClientM action env
+        queryCardanoWallet = Query (runQuery bailout)
+            (withAuthWallet mainPipeline lock bkToken)
+        queryAdmin = Query (runQuery bailout)
+            (withAuthWallet adminPipeline lock bkToken)
+    _checkPoint <- S.head_
+        $ getBuilds queryAdmin
+        & S.chain (traceWith logsTracer . ProgressAdmin . number)
+        & flip S.for (getArtifacts queryAdmin)
+        & S.filter ((==) "finished" . state . snd)
```

Comment:
are two  `S.filter` calls result in two "runs" or it is "merged into one call"?  Probably not a problem here but would be good to be aware how it behaves. btw. we have `beautiful folds approach` that do many things at once in ONE run - see for example https://hackage.haskell.org/package/folds and https://www.youtube.com/watch?v=6a5Ti0r8Q2s